File: video-01-syntax-basics/segment-05-loops/loops.py
# loops.py — Segment 05: Loops
#
# Python: range(), enumerate(), zip(), for/else — several things Java can't do.
# Java equivalent: Loops.java

print("=== Segment 05: Loops ===")

# --- range() replaces C-style for ---
print("range(5)   :", list(range(5)))          # range(stop)
print("step of 2  :", list(range(0, 10, 2)))   # range(start, stop, step)
print("countdown  :", list(range(5, 0, -1)))   # reverse — no extra variable needed

# --- for-each is just "for" ---
services = ["auth-service", "order-service", "payment-service"]
print("foreach    :", " ".join(services))

# --- enumerate(): index + value — no manual counter needed ---
print("indexed    :")
for i, service in enumerate(services):         # cf. Java: int i + services.get(i)
    print(f"  {i}: {service}")

# --- zip(): iterate two lists together ---
ports = [8080, 8081, 8082]
print("zipped     :")
for service, port in zip(services, ports):
    print(f"  {service} -> :{port}")

# --- while loop ---
n = 1
powers = []
while n <= 32:
    powers.append(n)
    n *= 2
print("powers of 2:", powers)

# --- for / else — else runs ONLY if loop was NOT broken ---
# Java devs: there is NO equivalent; Java uses a boolean flag instead
for service in services:
    if service == "billing-service":
        print("Found billing-service!")
        break
else:
    print("billing-service not found (for/else -- no Java equivalent)")

# --- List comprehension preview ---
# cf. Java: IntStream.range(0,6).map(x->x*x).boxed().toList()
squares = [x ** 2 for x in range(6)]
print("squares    :", squares)
# Deep dive on comprehensions coming in Video 2

File: video-01-syntax-basics/segment-05-loops/Loops.java
// Loops.java — Segment 05: Loops
//
// Java: C-style for, enhanced-for, while. Index tracking is manual.
// Python equivalent: loops.py

import java.util.List;

public class Loops {

    public static void main(String[] args) {
        System.out.println("=== Segment 05: Loops ===");

        // --- C-style for (range equivalent) ---
        System.out.print("range(5)   : ");
        for (int i = 0; i < 5; i++) {
            System.out.print(i + " ");
        }
        System.out.println();

        // --- Counting with step ---
        System.out.print("step of 2  : ");
        for (int i = 0; i < 10; i += 2) {
            System.out.print(i + " ");
        }
        System.out.println();

        // --- Enhanced for-each ---
        var services = List.of("auth-service", "order-service", "payment-service");
        System.out.print("foreach    : ");
        for (String service : services) {
            System.out.print(service + " ");
        }
        System.out.println();

        // --- Index + value — must track index manually (no enumerate) ---
        System.out.println("indexed    :");
        for (int i = 0; i < services.size(); i++) {
            System.out.printf("  %d: %s%n", i, services.get(i));
        }

        // --- while loop ---
        int n = 1;
        System.out.print("powers of 2: ");
        while (n <= 32) { System.out.print(n + " "); n *= 2; }
        System.out.println();

        // --- No for/else — simulate with a boolean flag ---
        boolean found = false;
        for (String s : services) {
            if (s.equals("billing-service")) { found = true; break; }
        }
        if (!found) System.out.println("billing-service not found (flag pattern)");
    }
}
